### PR TITLE
improved unstake and claim methods

### PR DESCRIPTION
**Pull request by Messer4**

### CODE REVIEW

File: solana.js [hunk 1]
```diff
@@ -286,6 +337,162 @@ async function stake(token, sender, lamports, source) {
...
+
+/** unstake - unstake
+ * @param {string} token - auth API token
+ * @param {string} sender - account blockchain address (staker)
+ * @param {number} lamports - lamport amount
+ * @param {string} source - stake source
+ * @returns {Promise<object>} Promise object with Versioned Tx
+ */
+async function unstake(token, sender, lamports, source) {
+    // if (!await CheckToken(token)) {
```
Comment (Elemberg) [inline solana.js:349]: why is it commented ?

File: solana.js [hunk 2]
```diff
@@ -96,6 +98,47 @@ async function createWithSeedTx(authorityPublicKey, lamports, source) {
...
+            stakePubkey: oldStakeAccountPubkey,
+            authorizedPubkey: authorityPublicKey,
+            splitStakePubkey: newStakeAccountPubkey,
+            basePubkey: authorityPublicKey,
+            seed: seed,
+            lamports: lamports
+        })
+       );
+        
+    return [splitStakeAccountTx, newStakeAccountPubkey, []]
```
Comment (Elemberg) [inline solana.js:129]: why we need 3rd value in this tuple ?

File: solana.js [hunk 3]
```diff
@@ -286,6 +337,162 @@ async function stake(token, sender, lamports, source) {
...
+ * @param {number} lamports - lamport amount
+ * @param {string} source - stake source
+ * @returns {Promise<object>} Promise object with Versioned Tx
+ */
+async function unstake(token, sender, lamports, source) {
+    // if (!await CheckToken(token)) {
+    //     throw new Error(ERROR_TEXT);
+    // }
+    
+    try {
```
Comment (Elemberg) [inline solana.js:353]: Do we need this try catch ?

File: solana.js [hunk 4]
```diff
@@ -286,6 +337,162 @@ async function stake(token, sender, lamports, source) {
...
+            if (a.account.data.info.stake.delegation.stake.eq(b.account.data.info.stake.delegation.stake)) {
+                return a;
+            }
+
+            return b;
+        });
+
+        let accountsToDeactivate = [];
+        let accountsToSplit = [];
+        var i = 0;
```
Comment (Elemberg) [inline solana.js:390]: (not sure) but mb use let instead of var a bit better

File: solana.js [hunk 5]
```diff
@@ -286,6 +337,162 @@ async function stake(token, sender, lamports, source) {
...
+        }
+
+        // cast instructions to correct JSON Serialization
+        instructions = instructions.map((instruction) => {
+            return new TransactionInstruction(instruction)
+        })
+        
+        const versionedTX = await prepareTransaction(instructions, senderPublicKey, []);
+
+        // await SetStats(token, 'unstake', lamports / LAMPORTS_PER_SOL, sender, versionedTX, chain);
```
Comment (Elemberg) [inline solana.js:438]: also no need ?

File: solana.js [hunk 6]
```diff
@@ -286,6 +337,162 @@ async function stake(token, sender, lamports, source) {
...
+            });
+            instructions.push(...withdrawTx.instructions);
+        }
+
+        // cast instructions to correct JSON Serialization
+        instructions = instructions.map((instruction) => {
+            return new TransactionInstruction(instruction)
+        })
+     
+        // await SetStats(token, 'claim', stakeBalance / LAMPORTS_PER_SOL, address, withdrawTx, chain);
```
Comment (Elemberg) [inline solana.js:487]: same

File: solana.js (hunk 2 above)
Comment (Messer4) [reply solana.js:129]: Transaction could have additional signers

File: solana.js (hunk 3 above)
Comment (Messer4) [reply solana.js:353]: Possible in future

**Review by Elemberg — APPROVED** [review]

**Review by Elemberg — APPROVED** [review]

**Pull request merged**